README.md
# bunyan-layout

Tools for logging in Bunyan JSON format (see See: https://github.com/trentm/node-bunyan)
in various logging frameworks for Java.

* [log4j2-bunyan](log4j2-bunyan/) - Log4j 2.x log layout.
* [log4j-bunyan](log4j-bunyan/) - Log4j 1.4.x log layout. 
* [java-logging-bunyan](java-logging-bunyan/) - Java Util Logging log formatter.
* [logback-bunyan](logback-bunyan/) - Logback classic log layout.

## Development

The project uses git-flow branch strategy, see
[introduction](http://nvie.com/posts/a-successful-git-branching-model/)
and the [git-flow tool](https://github.com/nvie/gitflow). Mainly all
development goes into development branch and via releases into master.

### Building

Complete build and testing is run with maven: `mvn clean install`

See the jenkins job at:
https://jenkins.sys.kth.se/view/Integral/job/bunyan-layout

Pre-built binaries are published into Maven central. Publishing is done
with `mvn -Prelease` but requires additional configuration for access
to OSS Sonatype and a GPG key for signing.

### Release process with git flow

```
git flow release start x.y.z
mvn versions:set -DnewVersion=x.y.z
 *do whatever testing and update of RELEASENOTES.md*
 *commit changes*

git flow release finish x.y.z
```

It is possible to publish the release branch `git flow release publish x.y.z` to the 
repository if it is to be shared between developers or used in some CI environment.

.


RELEASENOTES.md
# Releasenotes

## 2.6.0

* Fix for logback logger so nested stacktraces are also logged

## 2.5.0

* Remove redundant fields from log4j layout

## 2.4.0

* Use thread id for pid.

## 2.3.0

* Update to Java8 thread safe time and date constructs.

## 2.2.0

* Additional Logback layout for use with e.g., Spring boot.

## 2.1.0

New release where former bunyan-layout for log4j and log4j2 are built as
separate sub modules and java-logging-bunyan is added as an additional
module.


src/test/java/se/kth/infosys/log4j/LogTest.java
package se.kth.infosys.log4j;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.junit.Test;

public class LogTest {
    private static Logger LOG = LogManager.getLogger(LogTest.class);

    @Test
    public void printStuff() {
        LOG.trace("Trace log");
        LOG.debug("Debug log");
        LOG.info("Info log");
        LOG.warn("Warn log");
        LOG.error("Error log");
        LOG.fatal("Fatal log");

        try {
            throw new RuntimeException("A runtime exception");
        }
        catch (Exception e) {
            LOG.error("An exception", e);
        }
    }
}


log4j-bunyan/src/test/java/se/kth/infosys/log4j/LogTest.java
package se.kth.infosys.log4j;

import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.junit.Test;

public class LogTest {
    private static Logger LOG = LogManager.getLogger(LogTest.class);

    @Test
    public void printStuff() {
        LOG.trace("Trace log");
        LOG.debug("Debug log");
        LOG.info("Info log");
        LOG.warn("Warn log");
        LOG.error("Error log");
        LOG.fatal("Fatal log");

        try {
            throw new RuntimeException("A runtime exception");
        }
        catch (Exception e) {
            LOG.error("An exception", e);
        }
    }
}


logback-bunyan/src/test/java/se/kth/infosys/logback/LogTest.java
package se.kth.infosys.logback;

import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class LogTest {
    private static Logger LOG = LoggerFactory.getLogger(LogTest.class);

    @Test
    public void printStuff() {
        LOG.trace("Trace log");
        LOG.debug("Debug log");
        LOG.info("Info log");
        LOG.warn("Warn log");
        LOG.error("Error log");

        try {
            throw new RuntimeException("A runtime exception", nestedException1());
        }
        catch (Exception e) {
            LOG.error("An exception", e);
        }
    }

    private RuntimeException nestedException1() {
        try {
            nestedException2();
        } catch (RuntimeException e) {
            return new RuntimeException("nested1", e);
        }
        return null;
    }

    private RuntimeException nestedException2() {
        throw new RuntimeException("nested2");
    }
}


logback-bunyan/src/main/java/se/kth/infosys/logback/BunyanLayout.java
package se.kth.infosys.logback;

import static java.util.Collections.singletonList;

import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.pattern.RootCauseFirstThrowableProxyConverter;
import ch.qos.logback.classic.pattern.ThrowableProxyConverter;
import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.classic.spi.IThrowableProxy;
import ch.qos.logback.core.LayoutBase;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.Map;

public class BunyanLayout extends LayoutBase<ILoggingEvent> {

    private static final Map<Level, Integer> BUNYAN_LEVEL;
    static {
        BUNYAN_LEVEL = new HashMap<>();
        BUNYAN_LEVEL.put(Level.ERROR, 50);
        BUNYAN_LEVEL.put(Level.WARN, 40);
        BUNYAN_LEVEL.put(Level.INFO, 30);
        BUNYAN_LEVEL.put(Level.DEBUG, 20);
        BUNYAN_LEVEL.put(Level.TRACE, 10);
    }

    private static final Gson GSON = new GsonBuilder().create();
    private ThrowableProxyConverter rtpc = new RootCauseFirstThrowableProxyConverter();

    @Override
    public void start() {
        rtpc.setOptionList(singletonList("full"));
        rtpc.start();
        super.start();
    }

    @Override
    public String doLayout(ILoggingEvent event) {
        JsonObject jsonEvent = new JsonObject();
        jsonEvent.addProperty("v", 0);
        jsonEvent.addProperty("level", BUNYAN_LEVEL.get(event.getLevel()));
        jsonEvent.addProperty("levelStr", event.getLevel().toString());
        jsonEvent.addProperty("name", event.getLoggerName());
        try {
            jsonEvent.addProperty("hostname", InetAddress.getLocalHost().getHostName());
        } catch (UnknownHostException e) {
            jsonEvent.addProperty("hostname", "unkown");
        }
        jsonEvent.addProperty("pid", getThreadId(event));
        jsonEvent.addProperty("time", formatAsIsoUTCDateTime(event.getTimeStamp()));
        jsonEvent.addProperty("msg", event.getFormattedMessage());

        if (event.getLevel().isGreaterOrEqual(Level.ERROR) && event.getThrowableProxy() != null) {
            JsonObject jsonError = new JsonObject();
            IThrowableProxy e = event.getThrowableProxy();

            jsonError.addProperty("message", e.getMessage());
            jsonError.addProperty("name", e.getClassName());
            jsonError.addProperty("stack", rtpc.convert(event));
            jsonEvent.add("err", jsonError);
        }
        return GSON.toJson(jsonEvent) + "\n";    }

    private long getThreadId(ILoggingEvent event) {
        long threadId;
        String threadName = event.getThreadName();
        if(Thread.currentThread().getName().equals(threadName)){
            threadId = Thread.currentThread().getId();
        } else {
            try {
                threadId = Long.parseLong(threadName.substring(threadName.lastIndexOf('-')));
            } catch (NumberFormatException e) {
                threadId = 0;
            }
        }
        return threadId;
    }

    private static String formatAsIsoUTCDateTime(long timeStamp) {
        final Instant instant = Instant.ofEpochMilli(timeStamp);
        return ZonedDateTime.ofInstant(instant, ZoneOffset.UTC)
                .format(DateTimeFormatter.ISO_INSTANT);
    }


}


java-logging-bunyan/src/main/java/se/kth/infosys/logging/BunyanFormatter.java
/*
 * MIT License
 *
 * Copyright (c) 2018 Kungliga Tekniska högskolan
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package se.kth.infosys.logging;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.time.Instant;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Formatter;
import java.util.logging.Level;
import java.util.logging.LogRecord;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;

/**
 * A java.util.logging.Formatter which formats logs in node-bunyan JSON format.
 *
 * Use it e.g. by setting the property:
 * java.util.logging.ConsoleHandler.formatter = se.kth.infosys.logging.BunyanFormatter
 *
 * See https://github.com/trentm/node-bunyan for details about log format.
 */
public class BunyanFormatter extends Formatter {
    private static final Gson GSON = new GsonBuilder().create();

    private static final ZoneId SYSTEM_ZONE = ZoneId.systemDefault();
    private static final Map<Level, Integer> BUNYAN_LEVEL;
    static {
        BUNYAN_LEVEL = new HashMap<Level, Integer>();
        BUNYAN_LEVEL.put(Level.SEVERE, 50);
        BUNYAN_LEVEL.put(Level.WARNING, 40);
        BUNYAN_LEVEL.put(Level.CONFIG, 30);
        BUNYAN_LEVEL.put(Level.INFO, 30);
        BUNYAN_LEVEL.put(Level.FINE, 20);
        BUNYAN_LEVEL.put(Level.FINER, 10);
        BUNYAN_LEVEL.put(Level.FINEST, 10);
    }

    private static String HOSTNAME;
    static {
        try {
            HOSTNAME = InetAddress.getLocalHost().getHostName();
        } catch (UnknownHostException e) {
            HOSTNAME = "unknown";
        }
    }

    private static String formatAsIsoUTCDateTime(long timeStamp) {
        final Instant instant = Instant.ofEpochMilli(timeStamp);
        return ZonedDateTime.ofInstant(instant, ZoneOffset.UTC)
                .format(DateTimeFormatter.ISO_INSTANT);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String format(LogRecord record) {
        final JsonObject jsonEvent = new JsonObject();

        jsonEvent.addProperty("v", 0);
        jsonEvent.addProperty("level", BUNYAN_LEVEL.get(record.getLevel()));
        jsonEvent.addProperty("name", record.getLoggerName());
        jsonEvent.addProperty("hostname", HOSTNAME);
        jsonEvent.addProperty("pid", record.getThreadID());
        jsonEvent.addProperty("time", formatAsIsoUTCDateTime(record.getMillis()));
        jsonEvent.addProperty("msg", record.getMessage());
        jsonEvent.addProperty("src", record.getSourceClassName());

        final Throwable thrown = record.getThrown();
        if (thrown != null && record.getLevel().intValue() >= Level.WARNING.intValue()) {
            JsonObject jsonError = new JsonObject();

            jsonError.addProperty("message", thrown.getMessage());
            jsonError.addProperty("name", thrown.getClass().getSimpleName());

            StringWriter sw = new StringWriter();
            PrintWriter pw = new PrintWriter(sw);
            thrown.printStackTrace(pw);
            jsonError.addProperty("stack", sw.toString());
            jsonEvent.add("err", jsonError);
        }

        return GSON.toJson(jsonEvent) + "\n";
    }
}


src/main/java/se/kth/infosys/log4j/BanyanLayout.java
package se.kth.infosys.log4j;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.TimeZone;

import org.apache.log4j.Layout;
import org.apache.log4j.Level;
import org.apache.log4j.spi.LoggingEvent;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;

public class BanyanLayout extends Layout {
	private static final Map<Level, Integer> BANYAN_LEVEL;
	static {
		BANYAN_LEVEL = new HashMap<Level, Integer>();
		BANYAN_LEVEL.put(Level.FATAL, 60);
		BANYAN_LEVEL.put(Level.ERROR, 50);
		BANYAN_LEVEL.put(Level.WARN, 40);
		BANYAN_LEVEL.put(Level.INFO, 30);
		BANYAN_LEVEL.put(Level.DEBUG, 20);
		BANYAN_LEVEL.put(Level.TRACE, 10);
	}

	private static final TimeZone TZ = TimeZone.getTimeZone("UTC");
	private static final DateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm'Z'");
	static {
		DATE_FORMAT.setTimeZone(TZ);
	}
	private static final Gson GSON = new GsonBuilder().create();
	
	/**
	 * Format the event as a Banyan style JSON object.
	 */
	public String format(LoggingEvent event) {
		JsonObject jsonEvent = new JsonObject();
		jsonEvent.addProperty("v", 0);
		jsonEvent.addProperty("level", BANYAN_LEVEL.get(event.getLevel()));
		jsonEvent.addProperty("levelStr", event.getLevel().toString());
		jsonEvent.addProperty("logger", event.getLoggerName());
		try {
			jsonEvent.addProperty("hostname", InetAddress.getLocalHost().getHostName());
		} catch (UnknownHostException e) {
			jsonEvent.addProperty("hostname", "unkown");
		}
		jsonEvent.addProperty("pid", event.getThreadName());
		jsonEvent.addProperty("time", getTime(event.getTimeStamp()));
		jsonEvent.addProperty("msg", event.getMessage().toString());
		jsonEvent.addProperty("src", event.getLocationInformation().getClassName());

		if (event.getLevel().isGreaterOrEqual(Level.ERROR) && event.getThrowableInformation() != null) {
			JsonObject jsonError = new JsonObject();
			Throwable e = event.getThrowableInformation().getThrowable();
			
			jsonError.addProperty("message", e.getMessage());
			jsonError.addProperty("name", e.getClass().getSimpleName());
			
			StringWriter sw = new StringWriter();
			PrintWriter pw = new PrintWriter(sw);
			e.printStackTrace(pw);
			jsonError.addProperty("stack", sw.toString());
			jsonEvent.add("err", jsonError);
		}
		return GSON.toJson(jsonEvent) + "\n";
	}
	
	private String getTime(long timeStamp) {
		return DATE_FORMAT.format(new Date(timeStamp));
	}

	/**
	 * The throwable object is rendered in the output.
	 */
	public boolean ignoresThrowable() {
		return false;
	}

	@Override
	public String getContentType() {
		return "application/json";
	}

	public void activateOptions() {}
}
